MINOR: Fix incorrect pattern matching on `version` in `CheckpointFile`

ijuma opened the pull request:
Also add test and refactor things a little to make testing easier.

ijuma committed:
MINOR: Fix incorrect pattern matching on `version` in `CheckpointFile`

Also add test and refactor things a little to make testing easier.

ijuma commented:
Review by @junrao and @benstopford.

ijuma commented on core/src/main/scala/kafka/server/checkpoints/CheckpointFile.scala:
```diff
@@ -82,9 +82,8 @@ class CheckpointFile[T](val file: File, version: Int, formatter: CheckpointFileF
-        version match {
-          case version =>
+        line.toInt match {
+          case `version` =>
```
Without this change, it would always evaluate to true even if it the read version was not the same as the field `version`.

benstopford replied:
Wow. that's been there for years. Good spot.

ijuma replied:
The previous version was correct because it matched against CurrentVersion. The backticks are only necessary if the variable name starts with lowercase. A bit magical, I know.

junrao replied:
Thanks for catching this. Could we add a comment on when `` is needed?

ijuma replied:
Sure, done.

benstopford commented:
@ijuma  Thanks for the patch.  Refactor looks good. LGTM.

junrao commented:
@ijuma : Thanks for the patch. LGTM. Just a minor comment.

ijuma committed:
Add comment explaining usage of backticks

junrao commented on core/src/main/scala/kafka/server/checkpoints/CheckpointFile.scala:
```diff
@@ -83,6 +83,8 @@ class CheckpointFile[T](val file: File, version: Int, formatter: CheckpointFileF
           return Seq.empty
         line.toInt match {
+          // Backticks are required to match against the value of the version field instead of introducing a new
+          // variable called version
```
Oh, in that case, would it be clearer to just rename version in the constructor to sth like expectedVersion and then match to version here?

ijuma replied:
@junrao That doesn't quite work. The idea is that if you want to match on the value of a variable, it has to start with uppercase or backticks are required. As you say, this is not too clear for people who don't know though, so I pushed a new version that uses a guard instead. That should be clear and we no longer need a multi-line comment. Let me know if that's OK.

ijuma committed:
Use guard instead of matching on variable

This is clearer for people who are not familiar
with the details of matching on variables.

junrao commented:
@ijuma : Thanks for the improvement. That's much clearer. LGTM. I will let you merge it.
